lib/src/git.rs

```diff
@@ -757,3 +757,68 @@ pub struct Progress {
     pub bytes_downloaded: Option<u64>,
     pub overall: f32,
 }
+
+#[derive(Default)]
+struct PartialSubmoduleConfig {
+    path: Option<String>,
+    url: Option<String>,
+}
+
+/// Represents configuration from a submodule, e.g. in .gitmodules
+/// This doesn't include all possible fields, only the ones we care about
+#[derive(Debug, PartialEq)]
+pub struct SubmoduleConfig {
+    pub name: String,
+    pub path: String,
+    pub url: String,
+}
+
+pub fn parse_gitmodules(
+    gitmodules_path: &Path,
+) -> Result<BTreeMap<String, SubmoduleConfig>, git2::Error> {
+    let mut map: BTreeMap<String, PartialSubmoduleConfig> = BTreeMap::new();
+    let mut ret: BTreeMap<String, SubmoduleConfig> = BTreeMap::new();
+    let git_config = git2::Config::open(&gitmodules_path)?;
+
+    let entries = git_config.entries(Some("submodule\\..+\\."))?;
+    entries.for_each(|entry| {
+        if entry.name() == None || entry.value() == None {
+            return;
+        }
+        let (config_name, config_value) = match (entry.name(), entry.value()) {
+            (Some(name), Some(value)) => (name, value),
+            _ => return,
+        };
+
+        // config_name is of the form submodule.<name>.<variable>
+        let (submod_name, submod_var) = config_name
+            .strip_prefix("submodule.")
+            .unwrap()
+            .split_once(".")
+            .unwrap();
+
+        let map_entry = map
+            .entry(submod_name.to_string())
+            .or_insert(PartialSubmoduleConfig::default());
+
+        match submod_var {
+            "path" => map_entry.path = Some(config_value.to_string()),
+            "url" => map_entry.url = Some(config_value.to_string()),
+            _ => return,
+        };
+    })?;
+
+    map.iter_mut()
+        .filter(|(_, val)| val.path != None && val.url != None)
+        .for_each(|(key, val)| {
+            ret.insert(
+                key.clone(),
+                SubmoduleConfig {
+                    name: key.clone(),
+                    path: val.path.take().unwrap(),
+                    url: val.url.take().unwrap(),
+                },
+            );
+        });
```

Comment:
Nice! That's way cleaner.

Reply:
to avoid unwraps in such a scenario you can use `filter_map`
also doesn't into_iter mean you shouldn't need to clone keys?.

```rust
let res = map
    .into_iter()
    .filter_map(|(name, val)| match (val.path, val.url) {
        (Some(path), Some(url)) => SubmoduleConfig { name, path, url },
        _ => None,
    })
    .collect::<BTreeMap<_, _>>();
```

Reply:
or `Some((name.clone(), SubmoduleConfig { name, path: val.path?, url: val.url? })))` at your option.